pkg/skaffold/runner/runner.go

```diff
@@ -139,14 +154,79 @@ func NewForConfig(opts *config.SkaffoldOptions, cfg *latest.SkaffoldConfig) (*Sk
 			Monitor: monitor,
 			Trigger: trigger,
 		},
-		changeSet:            &changeSet{},
+		changeSet: &changeSet{
+			rebuildTracker: make(map[string]*latest.Artifact),
+			resyncTracker:  make(map[string]*sync.Item),
+		},
 		labellers:            labellers,
 		defaultLabeller:      defaultLabeller,
 		portForwardResources: cfg.PortForward,
 		imageList:            kubernetes.NewImageList(),
 		cache:                artifactCache,
 		runCtx:               runCtx,
-	}, nil
+		intentLock:           &gosync.Mutex{},
+		resetBuildIntent:     func() {},
+		resetSyncIntent:      func() {},
+		resetDeployIntent:    func() {},
+	}
+
+	r.setupTriggerCallbacks()
+
+	return r, nil
+}
+
+func (r *SkaffoldRunner) setupTriggerCallbacks() {
+	r.buildIntent = true
+	api := "api"
+	if strings.ToLower(r.runCtx.Opts.BuildTrigger) == api || r.runCtx.Opts.APIMode {
```

Comment:
please remove the duplication across these intent callbacks - I would want to make sure no accident copy paste errors change the locking logic. We can pass the intent name and the boolean pointer in for the three different scenarios as parameters for a generic version of these functions.

Reply:
updated, lmk what you think